generic/piranha-tree-sitter/src/piranha.rs

```diff
@@ -0,0 +1,534 @@
+      if self.rule_store.global_rules().len() == current_rules.len() {
+        break;
+      }
+    }
+  }
+
+  /// Gets all the files from the code base that (i) have the language appropriate file extension, and (ii) contains the grep pattern.
+  /// Note that `WalkDir` traverses the directory with parallelism.
+  pub fn get_files_containing_feature_flag_api_usage(&self) -> HashMap<PathBuf, String> {
+    let pattern = self.get_grep_heuristics();
+    info!(
+      "{}",
+      format!("Searching for pattern {}", pattern.as_str()).green()
+    );
+    let files: HashMap<PathBuf, String> = WalkDir::new(&self.path_to_codebase)
+      // Walk over the entire code base
+      .into_iter()
+      // Ignore errors
+      .filter_map(|e| e.ok())
+      // Filter files with the desired extension
+      .filter(|de| {
+        de.path()
+          .extension()
+          .and_then(|e| {
+            e.to_str()
+              .filter(|x| x.eq(&self.rule_store.language_name()))
+          })
+          .is_some()
+      })
+      // Read the file
+      .map(|f| {
+        (
+          f.path().to_path_buf(),
+          read_file(&f.path().to_path_buf()).unwrap(),
+        )
+      })
+      // Filter the files containing the desired regex pattern
+      .filter(|x| pattern.is_match(x.1.as_str()))
+      .collect();
+    #[rustfmt::skip]
+    info!("{}", format!("Will parse and analyze {} files.", files.len()).green());
+    return files;
+  }
+
+  /// Instantiate Flag-cleaner
+  pub fn new(args: PiranhaArguments) -> Self {
+    let graph_rule_store = RuleStore::new(&args);
+    Self {
+      rule_store: graph_rule_store,
+      path_to_codebase: String::from(args.path_to_code_base()),
+      relevant_files: HashMap::new(),
+    }
+  }
+
+  /// To create the current set of global rules, certain substitutions were applied.
+  /// This method creates a regex pattern matching these substituted values.
+  ///
+  /// At the directory level, we would always look to perform global rules. However this is expensive because
+  /// it requires parsing each file. To overcome this, we apply this simple
+  /// heuristic to find the (upper bound) files that would match one of our current global rules.
+  /// This heuristic reduces the number of files to parse.
+  ///
+  pub fn get_grep_heuristics(&self) -> Regex {
+    let reg_x = self
+      .rule_store
+      .global_rules()
+      .iter()
+      .flat_map(|r| r.grep_heuristics())
+      .sorted()
+      //Remove duplicates
+      .dedup()
+      //FIXME: Dirty trick to remove true and false. Ideally, grep heuristic could be a field in itself for a rule.
+      // Since not all "holes" could be used as grep heuristic.
+      .filter(|x| !x.as_str().eq("true") && !x.as_str().eq("false"))
+      .join("|");
+    Regex::new(reg_x.as_str()).unwrap()
+  }
+}
+
+// Maintains the updated source code content and AST of the file
+#[derive(Clone)]
+pub struct SourceCodeUnit {
```

Comment:
not sure about the Rust style of coding. Does it make sense to move `SourceCodeUnit` and its impl into a separate file?